src/config.rs

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright 2022 Two Sigma Open Source, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//! Command line flags (to control which requests we respond to).
+
+use std::time::Duration;
+
+use clap::builder::BoolishValueParser;
+
+use super::protocol::RequestType;
+
+pub trait RequestTypeIgnorer {
+    fn should_ignore(&self, ty: &RequestType) -> bool;
+}
+
+#[derive(clap::Parser, Clone, Copy, Debug, Default)]
+pub struct Config {
```

Comment:
Someone with more macro skills than me could probably reduce the boilerplate here.

Reply:
We absolutely should do something less boiler-plate-y. I think it makes sense to keep a set of enabled/disabled request types.

```rust
impl FromStr for RequestType {
  ... 
}

impl Config {
  fn from_env() -> Result<Self> {
    let mut disabled: HashSet<RequestType> = ...'
    for (key, val) in std::env::vars() {
      if key.starts_with("NSNCD_IGNORE") && env_bool(val) {
         let type: RequestType = string.strip_prefix().parse().expect("this is psuedocode fix it");
         disabled.insert(type);
      }
    }
    return disabled;
  }
}

```

This way we could also decide to define groups of request types to enable/disable together as well. I'd defer to @geofft on which groups make sense (and we can also bechmark and figure out whether a HashSet or just a slice is faster for lookups once we're done).

Unless there's a crate you already know about that has a nice derive-macro for FromStr I will totally help you write some `macro_rules` nonsense for `RequestType` boilerplate.

Reply:
Yeah, I thought about something like HashSet, but I figured a match and an index into a struct was simplest and fastest. I think this gets us off the ground in terms of having control, and we can work on the ergonomics later.

I can see an argument for operation groups like `--ignore-users` or `--ignore-groups` or `--ignore-network`. 

There are crates out there that let you derive iterators for enums and maybe you could use those in a macro, but I got shook by MSRV tbqh.

Reply:
> Yeah, I thought about something like HashSet, but I figured a match and an index into a struct was simplest and fastest. I think this gets us off the ground in terms of having control, and we can work on the ergonomics later.

I think the HashSet thing is much easier to maintain than the current thing, but I'll defer to you unless we get a tiebreaker vote. I'm not really worried about it taking the extra day or two to get done.

The more I think about it, the more I think it might be easier to do groups than individual messages because there's less config to parse.

```rust
if std::env::var("NSNCD_ENABLE_GROUPS").is_ok() {
   enabled.extend(GROUPS_MESSAGES)
}

if std::env::var("NSNCD_ENABLED_NETWORK").is_ok() {
  enabled.extend(ADDRESS_MESSAGES);
}
```

If you do want to keep going down the road enabling/disabling individual request types, here's a macro [(playground link)](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=71672b0d8120b765e83e453b6b353062) to do `FromStr` for `RequestType`. We'd have to do a tiny bit of shenanigans when parsing the env variables because `stringify!` preserves case and case conversion isn't const.

```rust
macro_rules! request_type {
    ($($name:ident),*) => {
        #[derive(Debug)]
        pub enum RequestType {
            $(
            $name
            ),*
        }
        
        impl std::str::FromStr for RequestType {
            type Err = ();
            
            fn from_str(s: &str) -> Result<RequestType, ()> {
                match s {
                    $(
                     stringify!($name) => Ok(RequestType::$name),
                    )*
                    _ => Err(()),
                }
            }
        }
    }
}

request_type! {
    Foo,
    Bar,
    Baz
}

fn main() {
    let v: Result<RequestType, _> = "Foo".parse();
    
    println!("{v:?}");
}
```

Reply:
I'd personally prefer "groups of request types" that can be split, rather than individual requests.

As we could see in https://github.com/twosigma/nsncd/pull/26 and https://github.com/twosigma/nsncd/pull/49, there's often multiple request types that need to be handled to get a specific lookup to work.

Providing fine-grained levers that then cause more to be disabled, because that's how client-side detection logic works seems to be very confusing.

I guess you want this mostly to be able to disable host lookups, in case this is causing issues? I'd probably really limit it to only that, because we know having these disabled might be a valid "working" configuration.

Reply:
Yeah, thinking better about this the controls should mirror nsswitch.conf. Lemme take this back to the drawing board.

Reply:
Rad, I like this approach a lot more.